lib/syslog/pool/pool.go

```diff
@@ -0,0 +1,79 @@
+package pool
+
+import (
+	"io"
+)
+
+// A LimitedConnPool is a connection pool, with the property that
+// the total number of live connections is limited to the size
+// parameter passed to New.
+type LimitedConnPool struct {
+	conns  chan io.WriteCloser // Available connections in the pool
+	live   chan struct{}       // Keep a count of living connections (in our hands, or the client)
+	signal chan struct{}       // Used to wake up the connection producer
+	err    chan error          // Send dial errors back to the client
+}
+
+// New returns a new LimitedConnPool.
+func New(dial func() (io.WriteCloser, error), size int) (*LimitedConnPool, error) {
+	// Tentative first try - if this doesn't work, we assume it never will
+	// and fail to initialize.
+	w, err := dial()
+	if err != nil {
+		return nil, err
+	}
+
+	p := LimitedConnPool{
+		conns:  make(chan io.WriteCloser, size),
+		live:   make(chan struct{}, size),
+		signal: make(chan struct{}),
+		err:    make(chan error),
+	}
+
+	p.conns <- w
+	p.live <- struct{}{}
+
+	// keep p.conns populated
+	go func() {
+		for range p.signal {
+			for len(p.live) < size {
+				w, err := dial()
+				if err != nil {
+					p.err <- err
+					continue
+				}
+				p.conns <- w
+				p.live <- struct{}{}
+			}
+		}
+	}()
+
+	// kick off the producer
+	p.signal <- struct{}{}
+
+	return &p, nil
+}
+
+// Put returns a connection to the pool.
+func (p *LimitedConnPool) Put(w io.WriteCloser, dead bool) error {
+	if dead {
+		<-p.live               // decrement the live count
+		p.signal <- struct{}{} // signal the connection dialer
+		return w.Close()
+	}
+	p.conns <- w
+	return nil
+}
+
+// Get gets a connection from the pool.
+func (p *LimitedConnPool) Get() io.WriteCloser {
+	return <-p.conns
```

Comment:
What do you think about wrapping the `io.WriteCloser` so that the `Put` method gets called automatically when `Close` is called?

Reply:
I don't think we can, because when we create the wrapper we don't know what the value of `dead` will be. Unless maybe we export a new interface type like `WriteCloseDeader` with a `Dead() bool` method? `WriteCloseVitalityChecker`? :)

```go
func (p *LimitedConnPool) put(w io.WriteCloser, dead bool) error {
    ...
}

type autoPutter struct {
	w   io.WriteCloser
	put func() error
}

func (w autoPutter) Write(p []byte) (int, error) { return w.w.Write(p) }
func (w autoPutter) Close() error { return w.put() }

func (p *LimitedConnPool) Get() io.WriteCloser {
	return autoPutter{
		w:   <-p.conns,
		put: p.put(w, dead), // whoops
	}
}
```

Or did you have something else in mind?

Reply:
`WriterCloseCoroner`? :)

Reply:
The other option could be to take a second function, `dead() bool`, as an argument to `NewLimited`.

Reply:
I think the other issue here is that we would keep re-wrapping connections each time they pass through the pool, eventually resulting in unbounded recursion.

Reply:
You can know the value of `dead` by wrapping the `Write` method and checking for an error. When an error occurs writing we can safely assume that the connection is dead ;)